Init web-sdk go-feature-flag

thomaspoignant opened the pull request:
## This PR
- 1st version of the web-sdk for GO Feature Flag

### Related Issues
<!-- add here the GitHub issue that this PR resolves if applicable -->

Fixes https://github.com/thomaspoignant/go-feature-flag/issues/662

toddbaert commented on libs/providers/go-feature-flag-web/package.json:
```diff
@@ -0,0 +1,12 @@
+    "current-version": "echo $npm_package_version"
+  },
+  "peerDependencies": {
+    "@openfeature/web-sdk": "*"
```
@thomaspoignant we could consider now releasing a non-experimental web-sdk version now that the spec is merged, which would allow you to actually version this more appropriately.

thomaspoignant replied:
@toddbaert Yes, it would be even better if we have a final release.

toddbaert replied:
I think we can release a sub 1.0 web-sdk tomorrow if you'd like, so you can actually put a proper version here. cc @beeme1mr @lukas-reining @kinyoklion

lukas-reining replied:
@toddbaert That would be really good.

thomaspoignant replied:
@toddbaert Perfect, I will wait for the release anyway before releasing the provider.

toddbaert replied:
https://github.com/open-feature/js-sdk/pull/506#issuecomment-1652158329

toddbaert commented on libs/providers/go-feature-flag-web/README.md:
````diff
@@ -0,0 +1,83 @@
+The `FlagdWebProvider` receives events from flagd with changes. Combined with the event API in the web SDK, this allows for subscription to flag value changes in clients.
+
+```typescript
+client.addHandler(ProviderEvents.Ready, () => {
````
You may want to switch this to a flag value changed example to match the text above, plus it's a more interesting feature.

toddbaert commented on libs/providers/go-feature-flag-web/src/lib/go-feature-flag-web-provider.ts:
```diff
@@ -0,0 +1,269 @@
+  }
+
+  async initialize(context: EvaluationContext): Promise<void> {
+    await this.fetchAll(context);
+    await this.connectWebsocket();
```
Could you do these at the same time with a Promise.all? Do they need to go in order?

toddbaert commented on libs/providers/go-feature-flag-web/src/lib/go-feature-flag-web-provider.ts:
```diff
@@ -0,0 +1,269 @@
+    }
+  }
+
+  async initialize(context: EvaluationContext): Promise<void> {
```
I think you should start your provider in a `NOT_READY` state, and update it once the init is done. This is actually _needed_ in other SDKs, but we have a small bug in the one (IMO) where we initialize every time. This can lead to (in some situations) multiple inits of the same provider.

See [here](https://github.com/open-feature/js-sdk/issues/505).

toddbaert commented on libs/providers/go-feature-flag-web/src/lib/model.ts:
```diff
@@ -0,0 +1,91 @@
+  ErrorCode,
+  EvaluationContextValue,
+} from '@openfeature/js-sdk';
+import {FlagValue, ResolutionDetails} from "@openfeature/web-sdk";
```
```suggestion
import {FlagValue} from "@openfeature/web-sdk";
```

toddbaert commented on libs/providers/go-feature-flag-web/tsconfig.lib.json:
```diff
@@ -0,0 +1,13 @@
+{
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    "lib": ["ES2015", "DOM"],
```
I wonder if we need new generators for client components. I guess you added `DOM` manually?

thomaspoignant replied:
Yes, I copy/paste it from the `flagd-web` provider, but yes I agree we need a generator for web provider.

toddbaert commented on libs/providers/go-feature-flag-web/src/lib/go-feature-flag-web-provider.ts:
```diff
@@ -0,0 +1,269 @@
+
+  onClose(): Promise<void> {
+    this._websocket?.close(1000, "Closing GO Feature Flag provider");
+    return Promise.resolve(undefined);
```
```suggestion
    return Promise.resolve();
```

toddbaert commented on libs/providers/go-feature-flag-web/src/lib/go-feature-flag-web-provider.ts:
```diff
@@ -0,0 +1,269 @@
+    return Promise.resolve(undefined);
+  }
+
+  async onContextChange(_: EvaluationContext, newContext: EvaluationContext): Promise<void> {
+    this._logger?.debug(`new context provided: ${newContext}`);
+    await this.fetchAll(newContext);
+  }
```
```suggestion
  onContextChange(_: EvaluationContext, newContext: EvaluationContext): Promise<void> {
    this._logger?.debug(`new context provided: ${newContext}`);
    return this.fetchAll(newContext);
  }
```

I think you can just return this since it's returning a void Promise.

toddbaert commented on libs/providers/go-feature-flag-web/src/lib/go-feature-flag-web-provider.ts:
```diff
@@ -0,0 +1,269 @@
+  GoFeatureFlagWebProviderOptions,
+  GOFeatureFlagWebsocketResponse,
+} from "./model";
+import axios from "axios";
```
TOTALLY up to you, but I would recommend just using `fetch`. With a simple wrapper it will work pretty well. The only gotcha is I think by default it doesn't throw on non-200, but you can change that without much effort. It will keep you package size down at least, which is nice for client projects where every byte of JS counts.

toddbaert replied:
You could also totally do this as a new issue :shrug:

toddbaert reviewed (approved):
You are the go-sdk expert, so I of course approve! I did leave a lot of optional comments. The only one that really worries me is [this one](https://github.com/open-feature/js-sdk-contrib/pull/474/files#r1274012627) (which is also related to an SDK bug).

toddbaert commented on libs/providers/go-feature-flag-web/src/lib/go-feature-flag-web-provider.ts:
```diff
@@ -0,0 +1,269 @@
+   * connectWebsocket is starting the websocket and associate some handler
+   * to react if the state of the websocket change.
+   */
+  async connectWebsocket(): Promise<void> {
```
Very very cool to have WS support!

thomaspoignant commented on libs/providers/go-feature-flag-web/src/lib/go-feature-flag-web-provider.ts:
```diff
@@ -0,0 +1,269 @@
+      await this.fetchAll(OpenFeature.getContext(), flagsChanged);
+    }
+    this._websocket.onclose = async () => {
+      this._logger?.info(`Websocket closed, trying to reconnect`);
```
It could be a lot of things here, such as the connection is closed from the upstream, the server does not exist anymore because we have downscale etc ...

But I agree using `warn` here is better.

thomaspoignant commented:
@toddbaert I've addressed all your comments, including using `fetch` rather than `axios`.
If you don't mind doing another review it would be awesome.

I think we should wait the release of a stable SDK to be available before merging this PR.

toddbaert commented:
Sorry @thomaspoignant I will re-review this first thing tomorrow

thomaspoignant commented:
> Sorry @thomaspoignant I will re-review this first thing tomorrow

No rush, no worries

toddbaert commented on libs/providers/go-feature-flag-web/src/lib/go-feature-flag-web-provider.ts:
```diff
@@ -0,0 +1,341 @@
+        this.handleFetchErrors(err)
+        await new Promise((resolve) => setTimeout(resolve, delay));
+        delay *= this._retryDelayMultiplier;
+        this._logger?.info(`xxx ${delay} ms (${attempt}/${this._maxRetries}).`)
```
Is this "xxx" supposed to be here?

Something I like to do is use the classname as a prefix to my log messages: `${GoFeatureFlagWebProvider.name}: ....`

toddbaert commented on libs/providers/go-feature-flag-web/src/lib/go-feature-flag-web-provider.ts:
```diff
@@ -0,0 +1,341 @@
+    const init: RequestInit = {
+      method: "POST",
+      headers: {
+        'Content-Type': 'application/json',
```
You could extract the common headers to a template object, or even add them into a convenience method.

toddbaert reviewed (approved):
@thomaspoignant looks great to me! Only minor nits that are opinions. Please be careful resolving conflicts in the `.release-please-manifest` :sweat_smile: 

I plan on releasing our first non-experimental web-sdk on Monday, with this change: https://github.com/open-feature/js-sdk/pull/507#issuecomment-1654678449. I'm waiting until then to give people a few days to update the providers impacted by the change.

You can wait until then if you'd like to set a proper peer for this.

thomaspoignant commented on libs/providers/go-feature-flag-web/package.json:
```diff
@@ -0,0 +1,12 @@
+    "current-version": "echo $npm_package_version"
+  },
+  "peerDependencies": {
+    "@openfeature/web-sdk": "^0.4.0"
```
I've stick the version to the official release to come.

beeme1mr commented:
@toddbaert, it looks like the CI is failing because of flagd. I'm assuming the NX affecting command detected so many components because the root package json was modified, right? Either way, we should unblock Thomas.

thomaspoignant commented:
> Awesome, nice job @thomaspoignant! We should extend the playground to support client-side GOFF once this has been published.

Sure I will look at it as soon as we merge it.
